packages/effect/src/Effect.ts

```diff
@@ -13652,17 +13660,42 @@ export const Service: <Self = never>() => [Self] extends [never] ? MissingSelfGe
     const layerName = hasDeps ? "DefaultWithoutDependencies" : "Default"
     let layerCache: Layer.Layer.Any | undefined
     if ("effect" in maker) {
-      Object.defineProperty(TagClass, layerName, {
-        get(this: any) {
-          return layerCache ??= layer.fromEffect(TagClass, map(maker.effect, (_) => new this(_)))
-        }
-      })
+      const isFunction = typeof maker.effect === "function"
+
+      if (isFunction) {
+        Object.defineProperty(TagClass, layerName, {
+          get(this: any) {
+            return function (this: typeof TagClass) {
+              return layerCache ??= layer.fromEffect(TagClass, map(maker.effect.apply(null, arguments), (_) => new this(_)))
```

Comment:
This would return the same Layer for different args.

It also leads into the memoization issue. Potentially a MutableHashMap could be used to try to keep a stable reference per argument set, but it won't catch everything so not sure about this.